### ISSUE TITLE
Support to CORS API's

### ISSUE THREAD

**Issue by mnatanbrito** [bjowes/cypress-ntlm-auth#65]
I'm not sure if this is a bug or a feature request or a cry for help 😥 . My solution comprises two distinct applications running in a VM: a React frontend (let's say frontend.company.com) hosted in an IIS site consuming an API (NancyFx on top of Asp.Net Core) on a different domain (backend.company.com), also hosted in IIS. The backend requires NTLM authentication from the frontend. In my current test setup, I have configured **cypress-ntlm-auth**  with:

`cy.ntlm('http://backend.company.com', 'marcos.pacheco', 'password', 'DMZ');`

before actually starting the interactions. I'm running the tests in my dev machine (not inside the VMs). As this is a CORS request, I have configured my backend to allow request from the proxy created by the plugin (as I wasn't sure which port would be considered the remote address, so I ended up allowing both in my backend):

![image](https://user-images.githubusercontent.com/4418014/61897812-44b9fd80-aeee-11e9-9b83-8ad60ba574c1.png)

However, as soon as I hit the login button (which by its turn issues a request to the userInfo authenticated endpoint), the request is aborted with the message:

> Access to XMLHttpRequest at 'http://backend.company.com/api/userInfo' from origin 'http://frontend.company.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

The following message is logged to the console in debug mode:
![image](https://user-images.githubusercontent.com/4418014/61899428-cbbca500-aef1-11e9-99df-4f9d078f5544.png)

By inspecting the cypress browser console, I found out that neither of the configured origins were being considered the remote address for the request but the address _http://127.0.0.1:57015_ as follows:
![image](https://user-images.githubusercontent.com/4418014/61899884-cf9cf700-aef2-11e9-86f3-4a29b637cfa1.png)

So I altered my CORS configuration to allow requests from _http://127.0.0.1:57015_. But the error still persists. Does anyone know what am I missing here? I would love to get Cypress + NTLM + CORS working to be able to proceed with the tests.

I have already checked the VM local computer policies as instructed by #49;
I have also captured the traffic as suggested in #25 and it seems the credentials are successfully sent:
![image](https://user-images.githubusercontent.com/4418014/61902746-2c031500-aef9-11e9-826f-ddb769048c7e.png)

🆘

**Comment by bjowes** [comment]
Hi @mnatanbrito, I know from experience that CORS can be a bit frustrating until you get it right. First, CORS is enforced by the client, not by the server. I personally found this odd, but I have seen several cases where backends send requests to other backends and simply ignore the CORS headers - still works fine. So it is Chrome you need to please here. This means that the CORS config on the server should match the address in Chrome - the proxy address don't matter. 

If your frontend is served from frontend.company.com, Chrome should accept the request if frontend.company.com is returned as Access-Control-Allow-Origin. But - when you are developing the frontend, it is common that the frontend is running on localhost, meaning that the CORS check will fail unless localhost with the right port is in Access-Control-Allow-Origin.

To make things more complex, Chrome does not allow credentials to be passed to the backend if Access-Control-Allow-Origin is set to * (the naive implementation). It must be an explicit match. This shouldn't concern your tests with the plugin, since Chrome isn't passing the credentials (the plugin does), but when running the site with real users it will matter. In previous projects I have been involved with we resolved this with the [IsOriginAllowed function in .Net Core](https://docs.microsoft.com/sv-se/dotnet/api/microsoft.aspnetcore.cors.infrastructure.corspolicy.isoriginallowed?view=aspnetcore-2.2). With this function we could use string matching to match parts of the hostname, and accept localhost with any port in our development environment.

Hopefully these pointers will help you on the way, but looking at your issue there is another thing that needs to be resolved first. According to Chrome: 'No 'Access-Control-Allow-Origin' header is present on the requested resource.' This means your backend isn't returning the correct header at all. Please start by validating that CORS is activated in the backend.

**Comment by mnatanbrito** [comment]
Hi, @bjowes thank you for your quick reply. Well, I believe the correct CORS headers are returned from the backend. I just changed my frontend application running on localhost to point to my real backend and it worked. The only difference is that it doesn't prompt me for the credentials. Obviously, I set the origin _http://localhost:3000_ as an allowed origin in my server config. I used the `SetIsOriginAllowed` method:

` builder.SetIsOriginAllowed((origin) => origin.StartsWith("http://localhost", System.StringComparison.InvariantCultureIgnoreCase) || origin.StartsWith("http://127.0.0.1", System.StringComparison.InvariantCultureIgnoreCase));`

So I'm guessing it has more to do with the proxying mechanism (or maybe I'm just wrong here). This is a successful response from my application running on localhost:
![image](https://user-images.githubusercontent.com/4418014/61915489-0e46a780-af1b-11e9-88ab-e3181ccb3d9d.png)
And this is the request made by **cypress-ntlm-auth**:
![image](https://user-images.githubusercontent.com/4418014/61915627-caa06d80-af1b-11e9-9ad3-4ab509f807c5.png)

So, I'm guessing that **127.0.0.1:51802** is the actual proxy sitting in between the test environment and the real backend. Is it possible that the proxy is somehow not passing the right _Origin_ or _Referer_ to the backend? That would explain the backend not validating the preflight request from the test env through the proxy. I'm just trying out every possibility here 😞

**Comment by bjowes** [comment]
The proxy doesn’t change any incoming headers, so it shouldn’t remove the Origin header. 
What the proxy does though is to start an authentication handshake on all new connections. Normally the CORS preflight message should use Anonymous but the proxy will try to authenticate it too. In my own environments it works anyway, but maybe that is the actual issue here? Did you enable anonymous access to the site by the way? Otherwise the preflight won’t work from the browser. 

What you could try is to use wireshark to capture the traffic when running with the plugin. Check which headers are sent from the server to the proxy, and then from the proxy to the client. 

When you don’t get a login prompt in Chrome, Single sign on is active. It works out of the box if the server hostname is in the trusted sites (Internet Options in Windows). You probably don’t have local host in that list, and I’m a bit unsure if adding it would cause a security risk.

**Comment by mnatanbrito** [comment]
Once again thank you for your support! My backend has anonymous authentication enabled as follows:
![image](https://user-images.githubusercontent.com/4418014/61954424-366fee00-af8f-11e9-82c4-7a43051ceee9.png)
So I went on investigating the traffic between the parties involved. I filtered every request to or from my backend (which I hope would watch conversations with any proxy in between). I found inconsistencies between both requests. This is my real application pre-flight request:
![image](https://user-images.githubusercontent.com/4418014/61961523-efd5c000-af9d-11e9-8b4b-ea6898fd7f0f.png)
And this is the test environment pre-flight request:
![image](https://user-images.githubusercontent.com/4418014/61961621-26abd600-af9e-11e9-8f1a-c2343b91b610.png)
As you can see, it lacks CORS headers and sends the __Authorization__ header even before receiving the challenge from the backend. After this request, the server returns __401__ with no CORS headers which is different from the __204__ response with CORS headers I get when using the real application.
Do I need to run the tests inside the server in which both applications are hosted 🤔? (It's a dumb question I know, but I feel dumb right now so who cares 😅).

**Comment by bjowes** [comment]
No worries, I am happy to help! I usually learn something myself along the way...
Yes, it is correct that the proxy sends the Authorization header before the challenge. It was made this way as an optimisation (one less roundtrip per handshake), but I am starting to realise that I may have simplified things a bit too much. The thing is that CORS API's works fine in other environments I have tested it, but it seems that the authorization header messes up the preflight in your case. I think the proper solution is that I implement a full handshake - meaning don't touch the first request from the client and perform the handshake after receiving the first challenge. I have started working on this, I'll let you know when it is ready for testing.

**Comment by mnatanbrito** [comment]
Thanks again @bjowes 
Just let me know if I can help in any way. I would be happy to do it.

**Comment by bjowes** [comment]
Hi @mnatanbrito, I have just published a new beta version which implements a full NTLM handshake. This means it won't attempt the handshake on the pre-flight message, so CORS should work better in your setup. Please try it out! The beta also includes some improvements to the NTLMv2 implementation, but I don't think that affects your tests.
`npm install cypress-ntlm-auth@beta`

**Comment by gustavobmichel** [comment]
Hi @bjowes, I have a setup where I only allow Windows Authentication on IIS. When using http, everything is working as expected, but when I switch to https I start getting errors of CORS (as described by @mnatanbrito). I am only doing a get and I think that's why it's not doing the OPTIONS handshake.

I installed the newest beta you made available yesterday to check the issue, but I am not quite sure if maybe there's something missing on my CORS setup. HTTPS has been working fine for a while for our local users, so it leads me to believe my CORS setup has been somewhat correct. Based on your comment about the proxy not changing the headers, I am a bit confused on why I get the error using cypress.

Do I need to have Anonymous Authentication set ,even if I don't normally use it on any of my apps? Based on your comment on this thread I am assuming that it's a yes.

**Comment by mnatanbrito** [comment]
@bjowes Thank you for your work. I have tested the new version and the full handshake is in place. Now I can authenticate using the plugin 😄

**Comment by gustavobmichel** [comment]
@mnatanbrito Are you using HTTPS or HTTP? As soon as I change to HTTPS I start getting CORS errors  but with HTTP it's all good. I have checked my CORS setup and it seems to be fine.

**Comment by bjowes** [comment]
Hi @gustavobmichel, all the examples and logs from @mnatanbrito indicate HTTP, but it should work fine with HTTPS also. 

You say you are only using GET. This is required but not sufficient to avoid CORS. Please check first if you fulfil all the requirements for [simple requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Simple_requests).

When you say you switch to HTTPS, what do you switch? Frontend, backend or both? What are the addresses? Are you running from localhost or deployed? Please provide som logs - it will answer most of the above questions :)

**Comment by bjowes** [comment]
I'm closing this now as the original issue was resolved in a sharp release. @gustavobmichel we can keep discussing here but if we end up with a different issue with the plugin we'll open a new issue.
